File: EmarsysMobileShop/EmarsysMobileShop/RecommendationManager.m
//
//  RecommendedDataSourceAnddelegate.m
//  EmarsysMobileShop
//

#import "RecommendationManager.h"
#import "RecommendedItem.h"

@implementation RecommendationManager

static NSString *const reuseIdentifier = @"RecommendedItemCell";

- (NSInteger)numberOfSectionsInCollectionView:
    (UICollectionView *)collectionView {
    return 1;
}

- (NSInteger)collectionView:(UICollectionView *)collectionView
     numberOfItemsInSection:(NSInteger)section {
    return self.recommendResults.count;
}

- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView
                  cellForItemAtIndexPath:(NSIndexPath *)indexPath {

    RecommendedItem *cell =
        [collectionView dequeueReusableCellWithReuseIdentifier:reuseIdentifier
                                                  forIndexPath:indexPath];

    // Configure the cell

    NSAssert(cell, @"No cell");

    Item *item = self.recommendResults[indexPath.row];

    [cell.title setText:item.title];
    cell.item = item;
    NSURL *url = [NSURL URLWithString:item.image];

    NSURLSessionTask *task = [[NSURLSession sharedSession]
          dataTaskWithURL:url
        completionHandler:^(NSData *_Nullable data,
                            NSURLResponse *_Nullable response,
                            NSError *_Nullable error) {

          if (!data) {
              return;
          }

          UIImage *image = [UIImage imageWithData:data];
          if (!image) {
              return;
          }

          dispatch_async(dispatch_get_main_queue(), ^{
            RecommendedItem *updateCell =
                (id)[collectionView cellForItemAtIndexPath:indexPath];
            if (!updateCell) {
                return;
            }
            [updateCell.image setImage:image];
          });
        }];
    [task resume];

    return cell;
}

- (void)collectionView:(UICollectionView *)collectionView
    didSelectItemAtIndexPath:(NSIndexPath *)indexPath {
    if (self.delegate) {
        [self.delegate updateWithItem:self.recommendResults[indexPath.row]];
    }
}

@end

File: EmarsysMobileShopSwift/EmarsysMobileShopSwift/RecommendationManager.swift
//
//  RecommendationManager.swift
//  EmarsysMobileShopSwift
//
//

import UIKit

protocol RecommendationManagerDelegate {
    func updateWithItem(item: Item)
}

class RecommendationManager: NSObject, UICollectionViewDelegate, UICollectionViewDataSource {
    
    var recommendResults = [Item]()
    var delegate: RecommendationManagerDelegate? = nil
    
    func numberOfSectionsInCollectionView(collectionView: UICollectionView) -> Int {
        return 1
    }
    
    func collectionView(collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        return recommendResults.count
    }
    
    func collectionView(collectionView: UICollectionView, cellForItemAtIndexPath indexPath: NSIndexPath) -> UICollectionViewCell {
        let cell = collectionView.dequeueReusableCellWithReuseIdentifier("RecommendedItemCell", forIndexPath: indexPath) as! RecommendedItem
        let item = recommendResults[indexPath.row]
        cell.title.text = item.title
        cell.item = item
        
        let url = NSURL(string: item.image!)
        let session = NSURLSession.sharedSession()
        session.dataTaskWithURL(url!, completionHandler: {(data, response, error) in
            guard let data = data else { return }
            guard let image = UIImage(data: data) else { return }
            dispatch_async(dispatch_get_main_queue()) {
                guard let cell = collectionView.cellForItemAtIndexPath(indexPath) else {
                    NSLog("Failed to get cell for index path %@", indexPath)
                    return
                }
                let updateCell = cell as! RecommendedItem
                updateCell.image.image = image
            }
        }).resume()
        
        return cell
    }
    
    func collectionView(collectionView: UICollectionView, didSelectItemAtIndexPath indexPath: NSIndexPath) {
        guard let delegate = delegate else { return }
        delegate.updateWithItem(recommendResults[indexPath.row])
    }
}
